File: 1143-longest-common-subsequence/1143-longest-common-subsequence.py
class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        te1_len = len(text1)
        m, n = len(text1), len(text2)
        if m < n: return self.longestCommonSubsequence(text2, text1) # 길이가 2가 더 크다면
        dp = [0] * (n + 1)
        for i in range(1, m + 1):
            pre = 0
            for j in range(1, n + 1):
                cur = dp[j]
                dp[j] = pre + 1 if text1[i-1] == text2[j-1] else max(dp[j-1], cur)
                pre = cur
        return dp[-1]

File: 1143-longest-common-subsequence/1143-longest-common-subsequence.java
class Solution {
    public int longestCommonSubsequence(String text1, String text2) {
        int n = text1.length();
        int m = text2.length();
        
        if (m < n) {
            int tmp = n;
            n = m;
            m = tmp;
            
            String val = text1;
            text1 = text2;
            text2 = val;
        }
        
        int[] dp = new int[n + 1];
        int pre;
        int cur;
        
        for (int i = 1; i <= m; i++) {
            pre = 0;
            for (int j = 1; j <= n; j++) {
                cur = dp[j];
                if (text1.charAt(j-1) == text2.charAt(i-1)) {
                    dp[j] = pre + 1;
                } else {
                    dp[j] = Math.max(dp[j-1], cur);
                }
                pre = cur;
            }
        }
        
        return dp[n];
    }
}